### PR TITLE
 Allow Opacity to be set differently in both focused and unfocused terminals

### LINKED ISSUE #11092: Add support for Acrylic and Opacity in "Unfocused Appearance"
Please add "Acrylic" setting support to unfocused appearance options being introduced in PR #10317 

I'd really like to make my unfocused windows to have transparency instead of opaque. Ideally, I could make my unfocused windows more transparent than my focused window.

Right now, since unfocused windows are opaque, they appear more predominant on the desktop than the focused window appears when configured with acrylic enabled and semi-transparent.

### PR DESCRIPTION
**Pull request by Jaswir**
## Summary of the Pull Request
Closes #11092 

Allowing `opacity `to be set differently in both focused and unfocused terminals

## References and Relevant Issues
#11092 , references: #7158 

### Allowing Opacity to be set differently in both focused and unfocused terminals:
unfocused_opacity
image

jolsen

image

## `_runtimeFocusedOpacity`

Mike also had to say something about this: https://github.com/microsoft/terminal/issues/2531#issuecomment-1668442774

Initially I had something like `  _setOpacity(newAppearance->Opacity());`

But with the introduction of unfocused opacity we encounter new challenges:
When Adjusting the Opacity with **CTRL+SHIFT+Mouse Scroll Wheel** or **Set background opacity** in command pallette, the Runtime opacity changes, but when we go to unfocused and back to focused the opacity changes back to focused opacity in Settings. 

Also when adjusting opacity through the command palette the window becomes unfocused and then focused again after setting background opacity hence the  `  _setOpacity(newAppearance->Opacity());` would override the changes made through command palette
runtimeFocusedOpacity

command_pallette_focusswitches

With the introduction of unfocused opacity we encounter new challenges. The runtime opacity stores both the unfocused opacity and focused opacity from settings at different moments. This all works well until we combine this with Adjusting the Opacity with **CTRL+SHIFT+Mouse Scroll Wheel** or **Set background opacity** in command pallette. This brings the need for a separate Focused Opacity. When we change the runtime opacity with scroll wheel or through command pallette this value needs to be stored separately from the one in settings. So we can change back to it when going to unfocused mode and back to focused instead of the focused opacity defined in settings. 

## `skipUnfocusedOpacity` solves Opacity going from solid to unfocused to focused bug: 
skipUnfocusedOpacity_bug

## Validation Steps Performed

- Checked if unfocused Opacity works well when adjusting opacity through Mouse Scroll Wheel or Command Palette and in combination with Acrylic as mentioned in "Detailed Description of the Pull Request / Additional comments"

## PR Checklist

- [x] Closes #11092 
- [ ] Tests added/passed
- [x] Documentation updated
   - If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here:(https://github.com/MicrosoftDocs/terminal/pull/714)
- [ ] Schema updated (if necessary)

### CODE REVIEW

**Comment by Jaswir** [comment]
Would be handy to have tests, so many different permutations to check with Acrylic + Opacity + Mouse Scroll Wheel + Adjusting Opacity through Command Palette

**Comment by Jaswir** [comment]
@lhecker

Made a new PR!

> (In particular, there's not much reason to performance optimize UI code since almost none of it is "hot code". But knowing what's slow is still useful because it avoids making architectural designs that are inherently slow and can't be optimized in hindsight.)

What would you consider "hot code" in this project?

**Comment by Jaswir** [comment]
Ugh there's still some bugs .. One sec

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -869,21 +871,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             _renderEngine->SetRetroTerminalEffect(newAppearance->RetroTerminalEffect());
             _renderEngine->SetPixelShaderPath(newAppearance->PixelShaderPath());
 
+            // Skip Unfocused Opacity when EnableUnfocusedAcrylic is false
+            // and Acrylic is true as this results into seeing the opacity going
+            // from solid to unfocused to focused bug.
+            bool skipUnfocusedOpacity = !focused && !(!_settings->EnableUnfocusedAcrylic() && UseAcrylic());
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:877]:
```suggestion
            bool skipUnfocusedOpacity = !focused && (_settings->EnableUnfocusedAcrylic() || !UseAcrylic());
```
Personally, I'd prefer that you [De Morgan's law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) it haha

**Review by carlos-zamora — APPROVED** [review]
Looks and feels great! Tested it out on my machine even :)

Here's some other feedback:
- [ ] Please update [the docs repo](https://github.com/MicrosoftDocs/terminal)

I'll approve since the comments are pretty small, but please make sure you get to them. Thanks and great work!

**Comment by Jaswir** [comment]
@zadjii-msft 

Hey Mike,
How was your Hackathon week? Hope it was nice.

I was wondering if you got a chance yet to answer my question:
https://github.com/microsoft/terminal/issues/2531#issuecomment-1729301477

**Comment by Jaswir** [comment]
> Looks and feels great! Tested it out on my machine even :)
> 
> Here's some other feedback:
> 
> * [ ]  Please update [the docs repo](https://github.com/MicrosoftDocs/terminal)
> 
> I'll approve since the comments are pretty small, but please make sure you get to them. Thanks and great work!

Thanks Carlos. 

"Looks and feels great!"

xD, I am glad you say that. I actually used to be one of the best (1000) game designers in The Netherlands if I say so myself in 2018, 2019 and there's this concept called Game Feel which is a lot about how things feel. Usually needed lots of player testing before it's good. Glad you like it! Hope the users will like it as well. If you like Game Feel and want to see a fun presentation about it see this: https://www.youtube.com/watch?v=AJdEqssNZ-U

I'll look into the docs and comments. Not sure what to add to the docs, let me try something..

**Comment by Jaswir** [comment]
@zadjii-msft @carlos-zamora 

I think my expectations of you guys may be wrong. I expected Microsoft Engineers would like go over all the possible scenario's when reviewing. in this case try all the various permutations of UseAcrylic , Opacity in Unfocused and Focused mode + all other combinations with Adjust Opacity in Command Palette and Scroll Wheel, etc. Maybe come up with things I missed, like edgecases about the feature (not code stuff). But didn't notice this in 7158 or this one. 

It seems like your reviewing process is mostly aimed at the code quality?
What is the reviewing process of code like at Microsoft? I guess this differs per team?

**Comment by Jaswir** [comment]
I reworded the code this should be good: 
``` cpp
      bool useFocusedRuntimeOpacity = focused || (!_settings->EnableUnfocusedAcrylic() && UseAcrylic());
      double newOpacity = useFocusedRuntimeOpacity ?
                              FocusedOpacity() :
                              newAppearance->Opacity();
      _setOpacity(newOpacity, focused);
```
Having some trouble making the code clear to you / other developers for this project. 
Essentially there are 2 special things going on

**Comment by Jaswir** [comment]
@carlos-zamora @zadjii-msft 
Ok I reworded the conditionals and added a function description like with some of the other functions to describe what the new parameter does and adapted my comments to make things more clear for other developers hopefully. 

Let me know what you think!

**Comment by Jaswir** [comment]
@zadjii-msft 

Hi Mike, 

This might be a silly way to do this. 
I'll just come up straight with it.
I am looking for a referral for Microsoft.
Can you give me a referral?

Solving 7158 boosted my confidence to start seriously applying to Big Tech Jobs. Can you give me a referral? I rather have a referral from you then someone that doesn’t know me.

Also, are there any relevant opportunities at Microsoft that would let me work on this project or similar ones to it?

@DHowett @lhecker @carlos-zamora @j4james 
I am not sure how to approach this situation hence I put this in the group with you more Senior Experienced Engineers.

**Comment by DHowett** [comment]
@jaswir sorry for the silence! We are all heads down preparing for our release on Tuesday. We'll have some more mental bandwidth as a team for discussions and PR reviews after that 🙂

**Comment by Jaswir** [comment]
> @Jaswir sorry for the silence! We are all heads down preparing for our release on Tuesday. We'll have some more mental bandwidth as a team for discussions and PR reviews after that 🙂

Thanks for letting me know! Will unfocused Acrylic be in the release ???

**Comment by Jaswir** [comment]
@carlos-zamora  Updated the docs

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 1 above)
Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.cpp]: Should be good now

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 2]
```diff
@@ -869,21 +871,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             _renderEngine->SetRetroTerminalEffect(newAppearance->RetroTerminalEffect());
             _renderEngine->SetPixelShaderPath(newAppearance->PixelShaderPath());
 
+            // Skip Unfocused Opacity when EnableUnfocusedAcrylic is false
+            // and Acrylic is true as this results into seeing the opacity going
+            // from solid to unfocused to focused bug.
+            bool skipUnfocusedOpacity = !focused && !(!_settings->EnableUnfocusedAcrylic() && UseAcrylic());
+            double newOpacity = skipUnfocusedOpacity ? newAppearance->Opacity() : FocusedOpacity();
```
[the remark this answers is not in the document]
Comment (Jaswir) [reply src/cascadia/TerminalControl/ControlCore.cpp]: @zadjii-msft Should be good now

**Comment by Jaswir** [comment]
@zadjii-msft Did you get a chance to rereview it yet? It's blocking https://github.com/microsoft/terminal/pull/15717

**Review by zadjii-msft — APPROVED** [review]

**Pull request merged**